### PR TITLE
feat: Stats for datasource cache

### LINKED ISSUE #30420: Log datasource hit/miss stats per repo
### Describe the proposed change(s).

We need datasource results to return in indicator if a cached result was returned or not, and then at the end of a repo we should log the cached vs non-cached count plus the percentage (in semantic format so that it can be collected/graphed by logging systems).

### PR DESCRIPTION
**Pull request by zharinov**
## Context

- Closes: #30420

## Documentation (please check one with an [x])

- [ ] I have updated the documentation, or
- [x] No documentation update is required

## How I've tested my work (please select one)

I have verified these changes via:

- [ ] Code inspection only, or
- [ ] Newly added/modified unit tests, or
- [ ] No unit tests but ran on a real repository, or
- [x] Both unit tests + ran on a real repository

### CODE REVIEW

File: lib/util/stats.ts [hunk 1]
```diff
@@ -105,6 +105,128 @@ export class PackageCacheStats {
...
+interface DatasourceCacheDataPoint {
+  datasource: string;
+  registryUrl: string;
+  packageName: string;
+  action: 'hit' | 'miss' | 'set' | 'skip';
+}
+
+export interface DatasourceCacheReport {
+  long: {
+    [datasource in string]: {
```
Comment (viceice) [inline lib/util/stats.ts:117]: looks ugly, can we use `Record`?

Comment (zharinov) [reply lib/util/stats.ts:117]: It's ugly, but it better communicates the name/purpose of the particular level of nesting

**Comment by rarkins** [comment]
Does it log any messages yet?

File: lib/util/stats.ts (hunk 1 above)
Comment (viceice) [reply lib/util/stats.ts:117]: I still don't like that

**Review by viceice — APPROVED** [review]

**Review by rarkins — APPROVED** [review]

**Pull request merged**